packages/wrangler/src/inspect.ts

```diff
@@ -0,0 +1,526 @@
+    });
+
+    ws.addEventListener("close", () => {
+      clearInterval(keepAliveInterval);
+    });
+
+    return () => {
+      // clean up! Let's first stop the heartbeat interval
+      clearInterval(keepAliveInterval);
+      // Then we'll send a message to the devtools instance to
+      // tell it to clear the console.
+      wsServerRef.current.clients.forEach((client) => {
+        // We could've used `localSocket` here, but
+        // then we would have had to add it to the effect
+        // change detection array, which would have made a
+        // bunch of other stuff complicated. So we'll just
+        // cycle through all of the server's connected clients
+        // (in practice, there should only be one or zero) and send
+        // the Log.clear message.
+        client.send(
+          JSON.stringify({
+            // TODO: This doesn't actually work. Must fix.
+            method: "Log.clear",
+            // we can disable the next eslint warning since
+            // we're referencing a ref that stays alive
+            // eslint-disable-next-line react-hooks/exhaustive-deps
+            id: messageCounterRef.current++,
+            params: {},
+          })
+        );
+      });
+      // Finally, we'll close the websocket
+      close();
+      // And we'll clear `remoteWebsocket`
+      setRemoteWebSocket(undefined);
+    };
+  }, [
+    props.inspectorUrl,
+    retryRemoteWebSocketConnectionSigil,
+    props.logToTerminal,
+  ]);
+
+  // This effect tracks the state changes _between_ the local
+  // and remote websockets, and handles how messages flow between them.
+  useEffect(() => {
+    /**
+     * This event listener is used for buffering messages from
+     * the remote websocket, and flushing them
+     * when the local websocket connects.
+     */
+    function remoteMessageBufferListener(event: MessageEvent) {
+      if (typeof event.data === "string") {
+        if (!localWebSocket) {
+          messageBufferRef.current.push(event.data);
+          // TODO: maybe we should have a max limit on this?
+          // if so, we should be careful when removing messages
+          // from the front, because they could be critical for
+          // devtools (like execution context creation, etc)
+        }
+      } else {
+        // We should never get here, but who know is 2022...
+        console.error("unrecognised devtools event:", event);
+      }
+    }
+
+    if (remoteWebSocket && !localWebSocket) {
+      // The local websocket hasn't connected yet, so we'll
+      // buffer messages until it does.
+      remoteWebSocket.addEventListener("message", remoteMessageBufferListener);
+    }
+
+    /** Send a message from the local websocket to the remote websocket */
+    function localMessageListener(event: MessageEvent) {
+      try {
+        remoteWebSocket.send(event.data);
+      } catch (e) {
+        if (e.message !== "WebSocket is not open: readyState 0 (CONNECTING)") {
```

Comment:
The message is brittle. I highly dount it's part of the spec. Honestly maybe this can just be silent. When I add a logger for https://github.com/cloudflare/wrangler2/issues/230 I'll revisit this. Leaving this conversation unresolved so I can find it later.